[KEW-1650] Updated GBM rewards dispensing calculation.

ernestopskins opened the pull request.

porkchop commented on contracts/eosio.system/src/producer_pay.cpp:
```diff
@@ -72,10 +72,12 @@ namespace eosiosystem {
          auto new_tokens = static_cast<int64_t>( (continuous_rate * double(token_supply.amount) * double(usecs_since_last_fill)) / double(useconds_per_year) );
+         const uint64_t days_in_gbm_period = useconds_in_gbm_period / useconds_per_day;
+         const uint64_t days_remaining_in_gbm_period = (gbm_final_time - current_time_point()).count();
          // needs to be 1/2 Savings, 1/6 Producers, 1/6 Voters, 1/6 Genesis Block Member
```
Ahh - something was left out in the descrition. We would also want to change the continuous rate to 5%: static constexpr double continuous_rate = 0.04879; // 5% annual rate here: https://github.com/worldwide-asset-exchange/wax-system-contracts/blob/develop/contracts/eosio.system/include/eosio.system/eosio.system.hpp#L62

and correspondingly, divide by 5 for the to_voters variable in the line 78 below. We should adjust this comment accordingly and maybe say that "gbm receives a linearly deflating share over three years"

porkchop commented on contracts/eosio.system/src/producer_pay.cpp:
```diff
@@ -72,10 +72,12 @@ namespace eosiosystem {
          auto to_voters        = new_tokens / 6;
          auto to_per_block_pay = to_voters;
-         auto to_gbm           = to_voters;
+         auto to_gbm           = to_voters * 2 * (double(days_remaining_in_gbm_period) / double(days_in_gbm_period));
```
I think we might want to calculate this in terms of usec remaining/total in keeping with the burn logic: https://github.com/worldwide-asset-exchange/wax-system-contracts/blob/develop/contracts/eosio.system/src/delegate_bandwidth.cpp#L593.

porkchop commented on contracts/eosio.system/src/producer_pay.cpp:
```diff
@@ -72,10 +72,12 @@ namespace eosiosystem {
          auto to_per_block_pay = to_voters;
-         auto to_gbm           = to_voters;
+         auto to_gbm           = to_voters * 2 * (double(days_remaining_in_gbm_period) / double(days_in_gbm_period));
          auto to_savings       = new_tokens - (to_voters + to_per_block_pay + to_gbm);
```
Since the continuous rate does not account for the gbm tokens being added, we will have to add the to_gbm amount to new_tokens above this line

ernestopskins replied:
Let's see if this is what you mean, Aaron.

porkchop replied:
yessir this is it :)

ernestopskins commented:
> I just realized, we likely need an if(to_gbm > 0) check on line 93. Pretty sure we won't be able to send 0 WAX when the to_gbm amount goes to aero

True. Fixed and fixed the transaction comments.

porkchop commented on contracts/eosio.system/src/producer_pay.cpp:
```diff
@@ -71,12 +71,16 @@ namespace eosiosystem {
-         auto to_gbm           = to_voters;
+         auto to_savings       = new_tokens - (to_voters + to_per_block_pay);
+         auto to_gbm           = to_voters * 2 * (delta_time_usec / double(useconds_in_gbm_period));
          auto to_savings       = new_tokens - (to_voters + to_per_block_pay + to_gbm);
```
hmm I think this line needs to be deleted. The assignment in line 80 looks like the correct value

ernestopskins replied:
Sure!, I believe this is a residual of the merge :man_facepalming:. Thanks.
